File: agentic-aws-nvidia-demo/services/orchestrator/app/analytics.py
"""
Mixpanel analytics integration for StoryWeave AI backend.

Provides event tracking, user property management, and analytics utilities.
"""
import os
import time
from typing import Optional, Dict, Any
from datetime import datetime
import mixpanel
from .settings import settings
from .logger import logger

# Initialize Mixpanel client
_mixpanel_client: Optional[mixpanel.Mixpanel] = None


def init_mixpanel(token: Optional[str] = None):
    """
    Initialize Mixpanel client.
    
    Args:
        token: Mixpanel project token (defaults to MIXPANEL_TOKEN env var)
    """
    global _mixpanel_client
    
    token = token or os.getenv("MIXPANEL_TOKEN")
    
    if not token:
        logger.warning("Mixpanel token not provided. Analytics disabled.")
        return None
    
    try:
        _mixpanel_client = mixpanel.Mixpanel(token)
        logger.info("Mixpanel initialized successfully")
        return _mixpanel_client
    except Exception as e:
        logger.error(f"Failed to initialize Mixpanel: {e}")
        return None


def get_client() -> Optional[mixpanel.Mixpanel]:
    """Get Mixpanel client instance."""
    global _mixpanel_client
    
    if _mixpanel_client is None:
        init_mixpanel()
    
    return _mixpanel_client


def _track_event(
    distinct_id: str,
    event_name: str,
    properties: Optional[Dict[str, Any]] = None
):
    """
    Track an event in Mixpanel.
    
    Args:
        distinct_id: User identifier
        event_name: Name of the event
        properties: Event properties
    """
    client = get_client()
    if not client:
        return
    
    # Add default properties
    default_properties = {
        "timestamp": int(time.time()),
        "app_version": os.getenv("APP_VERSION", "1.0.0"),
        "environment": settings.ENVIRONMENT,
    }
    
    if properties:
        default_properties.update(properties)
    
    try:
        client.track(distinct_id, event_name, default_properties)
    except Exception as e:
        logger.error(f"Failed to track event {event_name}: {e}")


def _set_user_property(distinct_id: str, properties: Dict[str, Any]):
    """
    Set user properties in Mixpanel.
    
    Args:
        distinct_id: User identifier
        properties: User properties to set
    """
    client = get_client()
    if not client:
        return
    
    try:
        client.people_set(distinct_id, properties)
    except Exception as e:
        logger.error(f"Failed to set user properties: {e}")


def _increment_user_property(distinct_id: str, property_name: str, value: int = 1):
    """
    Increment a user property.
    
    Args:
        distinct_id: User identifier
        property_name: Property name to increment
        value: Amount to increment by
    """
    client = get_client()
    if not client:
        return
    
    try:
        client.people_increment(distinct_id, {property_name: value})
    except Exception as e:
        logger.error(f"Failed to increment property {property_name}: {e}")


# Event Tracking Functions

def track_user_registered(user_id: int, email: Optional[str] = None, source: Optional[str] = None):
    """
    Track user registration event.
    
    Args:
        user_id: User ID
        email: User email
        source: Registration source (optional)
    """
    distinct_id = str(user_id)
    
    properties = {
        "source": source or "unknown",
    }
    
    if email:
        properties["email_domain"] = email.split("@")[-1] if "@" in email else None
    
    _track_event(distinct_id, "user_registered", properties)
    
    # Set user properties
    user_properties = {
        "signup_date": datetime.now().isoformat(),
        "total_stories_created": 0,
        "last_active": datetime.now().isoformat(),
    }
    
    if email:
        user_properties["email"] = email
    
    _set_user_property(distinct_id, user_properties)


def track_story_created(
    user_id: int,
    story_id: int,
    genre: Optional[str] = None,
    length: Optional[str] = None,
    has_corpus: bool = False
):
    """
    Track story creation event.
    
    Args:
        user_id: User ID
        story_id: Story ID
        genre: Story genre
        length: Story length
        has_corpus: Whether corpus was used
    """
    distinct_id = str(user_id)
    
    properties = {
        "story_id": story_id,
        "genre": genre or "unknown",
        "length": length or "unknown",
        "has_corpus": has_corpus,
    }
    
    _track_event(distinct_id, "story_created", properties)
    
    # Increment total stories count
    _increment_user_property(distinct_id, "total_stories_created")
    
    # Update favorite genre (increment genre counter)
    if genre:
        _increment_user_property(distinct_id, f"genre_{genre.lower().replace(' ', '_')}_count")


def track_outline_generated(
    user_id: int,
    story_id: int,
    generation_time: float,
    beat_count: int,
    genre: Optional[str] = None
):
    """
    Track outline generation event.
    
    Args:
        user_id: User ID
        story_id: Story ID
        generation_time: Time taken to generate (seconds)
        beat_count: Number of beats in outline
        genre: Story genre
    """
    distinct_id = str(user_id)
    
    properties = {
        "story_id": story_id,
        "generation_time": round(generation_time, 2),
        "beat_count": beat_count,
        "genre": genre or "unknown",
    }
    
    _track_event(distinct_id, "outline_generated", properties)


def track_scene_expanded(
    user_id: int,
    story_id: int,
    beat_index: int,
    scene_length: int,
    generation_time: Optional[float] = None
):
    """
    Track scene expansion event.
    
    Args:
        user_id: User ID
        story_id: Story ID
        beat_index: Index of the beat
        scene_length: Length of generated scene (characters)
        generation_time: Time taken to generate (seconds)
    """
    distinct_id = str(user_id)
    
    properties = {
        "story_id": story_id,
        "beat_index": beat_index,
        "scene_length": scene_length,
    }
    
    if generation_time:
        properties["generation_time"] = round(generation_time, 2)
    
    _track_event(distinct_id, "scene_expanded", properties)


def track_story_exported(
    user_id: int,
    story_id: int,
    format: str,
    scene_count: int,
    total_length: Optional[int] = None
):
    """
    Track story export event.
    
    Args:
        user_id: User ID
        story_id: Story ID
        format: Export format (txt, pdf, etc.)
        scene_count: Number of scenes exported
        total_length: Total length of exported story
    """
    distinct_id = str(user_id)
    
    properties = {
        "story_id": story_id,
        "format": format,
        "scene_count": scene_count,
    }
    
    if total_length:
        properties["total_length"] = total_length
    
    _track_event(distinct_id, "story_exported", properties)


def track_error_occurred(
    user_id: Optional[int],
    error_type: str,
    endpoint: Optional[str] = None,
    error_message: Optional[str] = None,
    **kwargs
):
    """
    Track error occurrence.
    
    Args:
        user_id: User ID (optional)
        error_type: Type of error
        endpoint: API endpoint where error occurred
        error_message: Error message
        **kwargs: Additional error properties
    """
    distinct_id = str(user_id) if user_id else "anonymous"
    
    properties = {
        "error_type": error_type,
        "endpoint": endpoint or "unknown",
    }
    
    if error_message:
        properties["error_message"] = error_message[:200]  # Limit length
    
    properties.update(kwargs)
    
    _track_event(distinct_id, "error_occurred", properties)


def track_api_call(
    user_id: Optional[int],
    endpoint: str,
    method: str,
    response_time: float,
    status_code: int
):
    """
    Track API call performance.
    
    Args:
        user_id: User ID (optional)
        endpoint: API endpoint
        method: HTTP method
        response_time: Response time in seconds
        status_code: HTTP status code
    """
    distinct_id = str(user_id) if user_id else "anonymous"
    
    properties = {
        "endpoint": endpoint,
        "method": method,
        "response_time": round(response_time, 3),
        "status_code": status_code,
    }
    
    _track_event(distinct_id, "api_call", properties)


def update_user_last_active(user_id: int):
    """
    Update user's last active timestamp.
    
    Args:
        user_id: User ID
    """
    distinct_id = str(user_id)
    _set_user_property(distinct_id, {
        "last_active": datetime.now().isoformat()
    })


def set_user_subscription_tier(user_id: int, tier: str):
    """
    Set user subscription tier.
    
    Args:
        user_id: User ID
        tier: Subscription tier (free, pro, enterprise)
    """
    distinct_id = str(user_id)
    _set_user_property(distinct_id, {
        "subscription_tier": tier
    })


def set_user_favorite_genre(user_id: int, genre: str):
    """
    Update user's favorite genre.
    
    Args:
        user_id: User ID
        genre: Favorite genre
    """
    distinct_id = str(user_id)
    _set_user_property(distinct_id, {
        "favorite_genre": genre
    })


# Initialize on import
init_mixpanel()

File: frontend/src/lib/analytics.ts
/**
 * Mixpanel analytics integration for StoryWeave AI frontend.
 * 
 * Provides event tracking, user property management, and React hooks.
 */
import mixpanel from 'mixpanel-browser';

let isInitialized = false;
let analyticsEnabled = true;

// Get Mixpanel token from environment
const MIXPANEL_TOKEN = import.meta.env.VITE_MIXPANEL_TOKEN;
const APP_VERSION = import.meta.env.VITE_APP_VERSION || '1.0.0';

/**
 * Initialize Mixpanel.
 */
export function initMixpanel(token?: string, enabled: boolean = true): void {
  if (isInitialized) {
    console.warn('Mixpanel already initialized');
    return;
  }

  const mpToken = token || MIXPANEL_TOKEN;
  analyticsEnabled = enabled;

  if (!mpToken) {
    console.warn('Mixpanel token not provided. Analytics disabled.');
    analyticsEnabled = false;
    return;
  }

  if (!enabled) {
    console.log('Analytics disabled by user preference');
    analyticsEnabled = false;
    return;
  }

  try {
    mixpanel.init(mpToken, {
      debug: import.meta.env.DEV,
      track_pageview: true,
      persistence: 'localStorage',
      ignore_dnt: false, // Respect Do Not Track
    });
    isInitialized = true;
    console.log('Mixpanel initialized successfully');
  } catch (error) {
    console.error('Failed to initialize Mixpanel:', error);
    analyticsEnabled = false;
  }
}

/**
 * Enable or disable analytics.
 */
export function setAnalyticsEnabled(enabled: boolean): void {
  analyticsEnabled = enabled;
  if (enabled && !isInitialized) {
    initMixpanel();
  }
  // Store preference
  localStorage.setItem('analytics_enabled', enabled.toString());
}

/**
 * Check if analytics is enabled.
 */
export function isAnalyticsEnabled(): boolean {
  if (!analyticsEnabled) return false;
  // Check localStorage preference
  const stored = localStorage.getItem('analytics_enabled');
  if (stored !== null) {
    return stored === 'true';
  }
  return analyticsEnabled;
}

/**
 * Track an event with properties.
 */
export function track(
  eventName: string,
  properties?: Record<string, any>
): void {
  if (!isAnalyticsEnabled() || !isInitialized) {
    return;
  }

  try {
    const defaultProperties = {
      app_version: APP_VERSION,
      timestamp: new Date().toISOString(),
      page: window.location.pathname,
    };

    mixpanel.track(eventName, {
      ...defaultProperties,
      ...properties,
    });
  } catch (error) {
    console.error(`Failed to track event ${eventName}:`, error);
  }
}

/**
 * Identify a user.
 */
export function identify(userId: string, userProperties?: Record<string, any>): void {
  if (!isAnalyticsEnabled() || !isInitialized) {
    return;
  }

  try {
    mixpanel.identify(userId);
    if (userProperties) {
      mixpanel.people.set(userProperties);
    }
  } catch (error) {
    console.error('Failed to identify user:', error);
  }
}

/**
 * Reset user identity (on logout).
 */
export function reset(): void {
  if (!isInitialized) {
    return;
  }

  try {
    mixpanel.reset();
  } catch (error) {
    console.error('Failed to reset user:', error);
  }
}

/**
 * Set user properties.
 */
export function setUserProperties(properties: Record<string, any>): void {
  if (!isAnalyticsEnabled() || !isInitialized) {
    return;
  }

  try {
    mixpanel.people.set(properties);
  } catch (error) {
    console.error('Failed to set user properties:', error);
  }
}

/**
 * Increment user property.
 */
export function incrementUserProperty(property: string, value: number = 1): void {
  if (!isAnalyticsEnabled() || !isInitialized) {
    return;
  }

  try {
    mixpanel.people.increment(property, value);
  } catch (error) {
    console.error(`Failed to increment property ${property}:`, error);
  }
}

// Event Tracking Functions

/**
 * Track page view.
 */
export function trackPageView(pageName: string, properties?: Record<string, any>): void {
  track('page_viewed', {
    page_name: pageName,
    ...properties,
  });
}

/**
 * Track button click.
 */
export function trackButtonClick(
  buttonName: string,
  page?: string,
  properties?: Record<string, any>
): void {
  track('button_clicked', {
    button_name: buttonName,
    page: page || window.location.pathname,
    ...properties,
  });
}

/**
 * Track feature usage.
 */
export function trackFeatureUsed(
  featureName: string,
  duration?: number,
  properties?: Record<string, any>
): void {
  const eventProperties: Record<string, any> = {
    feature_name: featureName,
    ...properties,
  };

  if (duration !== undefined) {
    eventProperties.duration = duration;
  }

  track('feature_used', eventProperties);
}

/**
 * Track search performed.
 */
export function trackSearchPerformed(
  query: string,
  resultCount?: number,
  properties?: Record<string, any>
): void {
  track('search_performed', {
    query_length: query.length,
    result_count: resultCount,
    ...properties,
  });
}

/**
 * Track tutorial completion.
 */
export function trackTutorialCompleted(
  tutorialName: string,
  duration?: number,
  properties?: Record<string, any>
): void {
  const eventProperties: Record<string, any> = {
    tutorial_name: tutorialName,
    ...properties,
  };

  if (duration !== undefined) {
    eventProperties.duration = duration;
  }

  track('tutorial_completed', eventProperties);
}

/**
 * Track story creation (frontend).
 */
export function trackStoryCreated(
  storyId: number,
  genre?: string,
  properties?: Record<string, any>
): void {
  track('story_created', {
    story_id: storyId,
    genre: genre || 'unknown',
    ...properties,
  });
}

/**
 * Track outline generation (frontend).
 */
export function trackOutlineGenerated(
  storyId: number,
  generationTime: number,
  beatCount: number,
  properties?: Record<string, any>
): void {
  track('outline_generated', {
    story_id: storyId,
    generation_time: generationTime,
    beat_count: beatCount,
    ...properties,
  });
}

/**
 * Track error occurrence (frontend).
 */
export function trackError(
  errorType: string,
  errorMessage?: string,
  properties?: Record<string, any>
): void {
  track('error_occurred', {
    error_type: errorType,
    error_message: errorMessage?.substring(0, 200), // Limit length
    page: window.location.pathname,
    ...properties,
  });
}

/**
 * Track user registration (frontend).
 */
export function trackUserRegistered(
  userId: string,
  source?: string,
  properties?: Record<string, any>
): void {
  identify(userId);
  track('user_registered', {
    source: source || 'unknown',
    ...properties,
  });

  // Set initial user properties
  setUserProperties({
    signup_date: new Date().toISOString(),
    total_stories_created: 0,
    last_active: new Date().toISOString(),
  });
}

/**
 * Update user last active.
 */
export function updateLastActive(): void {
  setUserProperties({
    last_active: new Date().toISOString(),
  });
}

/**
 * Set user subscription tier.
 */
export function setSubscriptionTier(tier: string): void {
  setUserProperties({
    subscription_tier: tier,
  });
}

/**
 * Track custom event with automatic properties.
 */
export function trackCustom(
  eventName: string,
  properties?: Record<string, any>
): void {
  track(eventName, properties);
}

// Initialize on load (respecting user preferences)
if (typeof window !== 'undefined') {
  const consent = localStorage.getItem('analytics_consent');
  if (consent === 'true') {
    initMixpanel();
  }
}
